### PR TITLE
Left align component browser to the edited node

### PR DESCRIPTION
**Pull request by vitvakatu**
### Pull Request Description

[ci no changelog needed]

[Task link](https://www.pivotaltracker.com/story/show/181870555)

This PR changes the relative position of the edited node in such a way that it is left-aligned to the component browser window. This change reflects the most recent version of the [design doc](https://github.com/enso-org/design/blob/main/epics/component-browser/design.md#overview)

Screenshot 2022-08-08 at 19 15 47

As an additional change, the FRP implementation of the `Camera2d` was extended with a new output (screen dimensions) and fixed. With the old implementation, there was a possibility of panic at runtime because of non-exclusive borrows of `RefCell`. The FRP event for camera position was emited inside the scope with a mutable `RefCell` borrow. Any attempt to borrow the camera one more time (e.g., by calling one of the getters, such as `zoom()`) caused panic at runtime.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

### CODE REVIEW

File: app/gui/view/src/component_browser.rs [hunk 1]
```diff
@@ -112,7 +112,7 @@ impl component::Frp<Model> for Frp {
                 &size,
                 &gap,
                 |list_size, list_x, size, gap| {
-                    Vector2(*list_x - list_size.x / 6.0, - size.y / 2.0 - gap - NODE_HEIGHT / 2.0)
+                    Vector2(*list_x - list_size.x / 2.0, - size.y / 2.0 - gap - NODE_HEIGHT / 2.0)
```
Comment (kazcw) [inline app/gui/view/src/component_browser.rs]: The `*list_x` could just be `list_x` (it was already like that, but as long as we're touching it, why not fix it).

It's not obvious what this math does. Either named subexpressions, or a comment, or maybe named constants could help the reader understand this calculation.

**Review by kazcw — APPROVED** [review]

File: lib/rust/ensogl/core/src/display/camera/camera2d.rs [hunk 2]
```diff
@@ -415,7 +416,9 @@ impl Camera2d {
 impl Camera2d {
     /// Sets screen dimensions.
     pub fn set_screen(&self, width: f32, height: f32) {
-        self.data.borrow_mut().set_screen(width, height)
+        self.data.borrow_mut().set_screen(width, height);
+        let screen = self.data.borrow().screen;
+        self.data.borrow().frp.screen.emit(screen);
```
Comment (farmaazon) [inline lib/rust/ensogl/core/src/display/camera/camera2d.rs]: I think we still could have the problem with double borrow - the `data` structure is borrowed while emitting `screen` event. We should instead clone_ref the screen endpoint from borrowed data and then emit the event.

Also, we can do this with only one borrow, avoiding unnecessary borrow checking:
```rust
   let (endpoint, screen) = {
       let mut data = self.data.borrow_mut();
       data.set_screen(width, height);
       (data.frp.screen.clone_ref(), data.screen)
   } // etc
```

File: lib/rust/ensogl/core/src/display/camera/camera2d.rs [hunk 3]
```diff
@@ -425,7 +428,14 @@ impl Camera2d {
 
     /// Update all dirty camera parameters and compute updated view-projection matrix.
     pub fn update(&self, scene: &Scene) -> bool {
-        self.data.borrow_mut().update(scene)
+        let is_updated = self.data.borrow_mut().update(scene);
+        if is_updated {
+            let borrowed = self.data.borrow();
+            let frp = &borrowed.frp;
+            frp.position.emit(self.display_object.position());
+            frp.zoom.emit(borrowed.zoom);
```
Comment (farmaazon) [inline lib/rust/ensogl/core/src/display/camera/camera2d.rs]: Here also we still borrow data while emitting events.

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: app/gui/view/src/component_browser.rs (hunk 1 above)
Comment (vitvakatu) [reply app/gui/view/src/component_browser.rs]: Thanks, fixed. Also added more subexpressions.

File: lib/rust/ensogl/core/src/display/camera/camera2d.rs (hunk 2 above)
Comment (vitvakatu) [reply lib/rust/ensogl/core/src/display/camera/camera2d.rs]: Nice catch, fixed. I also added a small unit test that catches invalid implementations. (not all of them)

File: lib/rust/ensogl/core/src/display/camera/camera2d.rs (hunk 3 above)
Comment (vitvakatu) [reply lib/rust/ensogl/core/src/display/camera/camera2d.rs]: Fixed

File: lib/rust/ensogl/core/src/display/camera/camera2d.rs [hunk 4]
```diff
@@ -535,3 +543,37 @@ impl display::Object for Camera2d {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    /// A regression test checks whether handling the camera's FRP events does not cause panics
+    /// at runtime.
+    ///
+    /// If the events are emitted with the camera's internal [`RefCell`] lock held, any usage of
+    /// camera API in event handlers will cause a panic.
```
Comment (farmaazon) [inline lib/rust/ensogl/core/src/display/camera/camera2d.rs]: Technically, not _any_ usage of camera API, but only those requiring mutable borrow (if they require only immutable borrow, and the lock is also immutable, the RefCell borrow is fine).

So make sure, that we test with the API requiring a mutable borrow.

**Review by farmaazon — APPROVED** [review]

File: app/gui/view/src/component_browser.rs [hunk 5]
```diff
@@ -112,7 +112,11 @@ impl component::Frp<Model> for Frp {
                 &size,
                 &gap,
                 |list_size, list_x, size, gap| {
-                    Vector2(*list_x - list_size.x / 6.0, - size.y / 2.0 - gap - NODE_HEIGHT / 2.0)
+                    let half_width = list_size.x / 2.0;
+                    let left_x = list_x - half_width;
+                    let bottom_y = -size.y / 2.0;
+                    let half_node_height = NODE_HEIGHT / 2.0;
+                    Vector2(left_x, bottom_y - gap - half_node_height)
```
Comment (wdanilo) [inline app/gui/view/src/component_browser.rs]: Please refactor that to a well-named function telling what these things do (they align the searcher to the left side, right?). It's hard to understand that by looking at these expressions alone. If refactorign would be hard, add comment of what this code does.

File: lib/rust/ensogl/core/src/display/camera/camera2d.rs [hunk 6]
```diff
@@ -425,7 +431,19 @@ impl Camera2d {
...
+            let pos_endpoint = borrowed.frp.position.clone_ref();
+            let zoom_endpoint = borrowed.frp.zoom.clone_ref();
+            let zoom = borrowed.zoom;
+            (is_updated, pos_endpoint, zoom_endpoint, zoom)
+        };
+        if is_updated {
+            pos_endpoint.emit(self.display_object.position());
+            zoom_endpoint.emit(zoom);
+        }
+        is_updated
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/camera/camera2d.rs]: 1. This code duplicates the code in `CameraData`. There is no explanation why. Is the code in `CameraData` still used? If no, why its still there. If yes, why?
2. This code uses a lot of strange clone_refs inside - there is not explanation why.
3. What does "_endpoint" suffix mean? Does it mean the FRP endpoint? If so, this is like naming a variable "age_var".
4. There is something bad with this code in general. It is too complex. It should be refactored somehow, but now, it does a lot of operations that are not obvious to the reader.

**Review by wdanilo — CHANGES_REQUESTED** [review]

File: lib/rust/ensogl/core/src/display/camera/camera2d.rs (hunk 6 above)
Comment (farmaazon) [reply lib/rust/ensogl/core/src/display/camera/camera2d.rs]: 1. Line 436
2. See point 4.
3. But you see the name clash, don't you? Anyway, I disagree: It's not like naming the variable `_var`. It's more like suffixing with `_value` - sometimes it's obvious that the var keeps value, but sometimes it's a ref, a cell, or, well, an FRP endpoint.
4. It was for my request: we don't want to emit any events having anything borrowed. Thus, we must clone_ref the FRP endpoints. I don't see any simpler way for doing that, except repeating the borrow() multiple times.

File: lib/rust/ensogl/core/src/display/camera/camera2d.rs (hunk 4 above)
Comment (vitvakatu) [reply lib/rust/ensogl/core/src/display/camera/camera2d.rs]: In a previous version of the code, the `position` event was emitted with the mutable lock held, so even an immutable borrow in some other method caused panic. My initial fix used immutable borrow, and it had an issue for methods with mutable borrow.  That's why we need to check both mutable and immutable borrows in the test. I will slightly rephrase this though

File: app/gui/view/src/component_browser.rs (hunk 5 above)
Comment (vitvakatu) [reply app/gui/view/src/component_browser.rs]: > they align the searcher to the left side, right?

The other way around, the expression_input_position is aligned to the left side of the component browser. Refactored (also removed one unnecessary input variable).

File: lib/rust/ensogl/core/src/display/camera/camera2d.rs (hunk 6 above)
Comment (vitvakatu) [reply lib/rust/ensogl/core/src/display/camera/camera2d.rs]: I simplified the code by copying the `frp` structure, not individual fields.

Comment (wdanilo) [reply lib/rust/ensogl/core/src/display/camera/camera2d.rs]: 1. Oh, there it is!
2. -
3. The name clash doesn't matter here really. I was referring to var names. If there is a name clash, between var `age` and `age`, renaming it to `age_var` is not a good solution. So I don't think the name clash should really contribute to this point. However, the explanation with "value-based suffixes" is understandable and I was not looking at it this way. I still somehow don't like it, but with this explanation, it makes much more sense in my head now.
4. I was never told that we should not do it. The request to never emit events from borrowed code is 100% valid. I was originally referring to the complexity of the implementation in general. The new version is way cleaner. If we did not want to clone the whole frp network, we could instead create a method on it allowing us to clone a sub-structure of it. However, if this is the frp network created with the new macro, we can just clone the input or output ports instead (they all are enclosed in a single RC). Anyway, @farmaazon, I was referring to code complexity, not to logical complexity.

**Review by wdanilo — APPROVED** [review]

**Pull request merged**